pkg/controller/snapshotter/snapshotter.go

```diff
@@ -347,204 +345,310 @@ func (vs *volumeSnapshotter) deleteSnapshot(spec *v1.PersistentVolumeSpec, sourc
+			glog.Errorf("Failed to get volume plugin. %v", err)
+			return nil, nil, true, fmt.Errorf("Failed to get volume plugin to create snapshot %s", snapshotName)
+		}
+		// Check whether the real snapshot is already created by the plugin
+		glog.Infof("findSnapshot: find snapshot %s by tags %v.", snapshotName, tags)
+		snapshotDataSource, snapStatus, err = (*plugin).FindSnapshot(tags)
+		if err != nil {
+			// Don't bail out because the error could be "error requesting snapshot : Resource not found".
+			glog.Infof("Error occurred when looking for snapshot by metadata: %v", err)
+		}
+		if snapshotDataSource != nil && snapStatus != nil {
+			glog.Infof("findSnapshot: found snapshot %s.", snapshotName)
+			return snapshotDataSource, snapStatus, true, nil
+		}
+	}
+	return nil, nil, bFoundMeta, nil
+}
 
-			// Check whether the VolumeSnapshotData object is already created
-			snapshotDataObj := vs.getSnapshotDataFromSnapshotName(snapshotName)
+func (vs *volumeSnapshotter) createSnapshot(snapshotName string, snapshot *crdv1.VolumeSnapshot) error {
+	glog.Infof("createSnapshot: Try to find metadata, and find snapshot %s by metadata.", snapshotName)
+	var snapshotDataSource *crdv1.VolumeSnapshotDataSource = nil
+	var snapStatus *[]crdv1.VolumeSnapshotCondition = nil
+	var bFoundMeta bool = false
+	var err error = nil
+	snapshotDataSource, snapStatus, bFoundMeta, err = vs.findSnapshot(snapshotName, snapshot)
+	if err != nil {
+		return err
+	}
 
-			// VolumeSnapshotData object not created yet. Create it now as snapshot has already been created.
-			if snapshotDataObj == nil {
-				readyCondition := crdv1.VolumeSnapshotDataCondition{
-					Type:    crdv1.VolumeSnapshotDataConditionPending,
-					Status:  v1.ConditionTrue,
-					Message: "Snapshot data is being created",
-				}
-				snapName := "k8s-volume-snapshot-" + string(uuid.NewUUID())
-
-				snapshotData := &crdv1.VolumeSnapshotData{
-					Metadata: metav1.ObjectMeta{
-						Name: snapName,
-					},
-					Spec: crdv1.VolumeSnapshotDataSpec{
-						VolumeSnapshotRef: &v1.ObjectReference{
-							Kind: "VolumeSnapshot",
-							Name: snapshotName,
-						},
-						PersistentVolumeRef: &v1.ObjectReference{
-							Kind: "PersistentVolume",
-							Name: pvName,
-						},
-						VolumeSnapshotDataSource: *snapshotDataSource,
-					},
-					Status: crdv1.VolumeSnapshotDataStatus{
-						Conditions: []crdv1.VolumeSnapshotDataCondition{
-							readyCondition,
-						},
-					},
-				}
-				var result crdv1.VolumeSnapshotData
-				for i := 0; i < createVolumeSnapshotDataRetryCount; i++ {
-					err = vs.restClient.Post().
-						Resource(crdv1.VolumeSnapshotDataResourcePlural).
-						Namespace(v1.NamespaceDefault).
-						Body(snapshotData).
-						Do().Into(&result)
-					if err != nil {
-						// Re-Try it as errors writing to the API server are common
-						time.Sleep(createVolumeSnapshotDataInterval)
-					} else {
-						break
-					}
-				}
+	var tags *map[string]string = nil
+	if !bFoundMeta {
```

Comment:
If using getSnapshotStatus in syncSnapshot(), we don't need to check Meta exist here or not. We are sure that snapshot has not yet created when calling this function createSnapshot()

Reply:
Yes

Reply:
Actually there are still several possibilities after calling getSnapshotStatus:
1. Metadata not created.
2. Metadata created but Snapshot not create yet.
3. Other situations ......
So we still need to know whether metadata is already created or not.  If not, create it.

Reply:
when calling createSnapshot, it means that getSnapshotStatus() returns statusNew. In this status, we know that snapshot request has not yet issued by cloud provider. It is possible that metadata has been created before, but we still want to create the metadata again to use the new timestamp right before issuing snapshot request.

Reply:
Sure.